Simple optional path to jobs api

Dhivyaa21 opened the pull request:
@cdancy @martinda - This pull-request is for issue#21 

I tried to get the value of  `@PathParam(OPTIONAL_FOLDER_PATH_PARAM) String optionalFolderPath'  variable inside the filter method of ScrubNullFolderParam.java, I was not successful. I think the values of pathparam variables were resolved already in the filter method.

But I found an annotation in jclouds called `@ParamParser` [here](https://jclouds.apache.org/reference/javadoc/1.8.x/org/jclouds/rest/annotations/ParamParser.html) which can intercept the path parameter variable values of the request. I have an initial implementation of this feature. I do have live tests for this change and it works locally for me. Let me know what you guys think.

cdancy commented:
Good start. Yeah I totally forgot about ParamParser and in my opinion that is where ALL the logic surrounding this and scrubbing should go. If you want to combine the scrub filter functionality into your param parser that would be great. After that we can see how things look.

Also ... Make sure to upper case class names.

cdancy commented:
And If you combine things you should be able to remove the optional-folder-path constant we use everywhere. Let me know if you need any help with this.

Dhivyaa21 commented on src/main/java/com/cdancy/jenkins/rest/JenkinsUtils.java:
```diff
@@ -288,6 +291,36 @@ public static void setEnvironmentVariables(final Map<String, String> newEnvVars)
+     *         path reconstructed to match the jenkins URL style.
+     */
+    public static String amendFolderPath(String folderPath) {
+        Pattern pattern = Pattern.compile("(job/[^!@#$%^&*<>]+/*)+");
```
Agreed! 

If we append "job/" always, then the optionalFolderPath variable can only take simple folder path expressions like "folder1/folder2".

Dhivyaa21 commented on src/main/java/com/cdancy/jenkins/rest/parsers/optionalFolderPathParser.java:
```diff
@@ -0,0 +1,21 @@
+ * Turn the optionalFolderPath param to jenkins URL style if needed
+ */
+@Singleton
+public class optionalFolderPathParser implements Function<Object,String> {
```
Typo! My bad. :(

cdancy commented on src/main/java/com/cdancy/jenkins/rest/JenkinsUtils.java:
```diff
@@ -288,6 +291,36 @@ public static void setEnvironmentVariables(final Map<String, String> newEnvVars)
+            if(i < folderNames.length-1) {
+                amendedPath += "/";
+            }
+        }
```
I don't want to add another library dependency for just one method if it can be helped. Instead we can recreate the method in our utilities file if need be.

Dhivyaa21 commented:
@cdancy - I removed the OPTIONAL_FOLDER_PATH_PARAM from the jobsApi class, replaced a null path with an empty string but I could not replace the request's base path "/" in the paramparser.  May be we still need the ScrubNullFolderParam filter. Let me know what you think.

cdancy commented on src/main/java/com/cdancy/jenkins/rest/parsers/OptionalFolderPathParser.java:
```diff
@@ -0,0 +1,19 @@
+    public static final String EMPTY_STRING = "";
+    @Override
+    public String apply(Object optionalFolderPath) {
+        return (optionalFolderPath == null) ? EMPTY_STRING : JenkinsUtils.amendFolderPath(String.class.cast(optionalFolderPath));
```
Lets bring that logic into this parser until we have a use-case not to do so.

cdancy commented:
@Dhivyaa21 one minor comment to address but all-in-all it LGTM.

cdancy commented:
> but I could not replace the request's base path "/" in the paramparser

@Dhivyaa21 can you provide an example of what you mean?

Dhivyaa21 commented:
@cdancy The request path to creating a job is `/{optionalFolderPath}/createItem`. The paramparser replaces  `{optionalFolderPath}` with an empty string if it is null which leaves the path as `//createItem`. This is why the tests are failing. I am not getting how to get rid of that extra "/" in the optionalFolderPathParser class.

cdancy commented:
@Dhivyaa21 just checked out your branch and looking into things now

cdancy commented:
@Dhivyaa21 so I think what we can do is replace the `@Path` annotations with something like:
```
@Path("{optionalFolderPath}job/{name}/api/json") // note the missing forward-slash behind _job_
```
We can then change the param-parser to ensure there is a leading forward-slash appended to the end if we have content otherwise return an empty-string. This way jclouds will be happy. We'll have to relax a bit on the rules we set which I think is fine. If devs want to append a forward slash to the end of their optional-path-param I think allowing it is fine for the sake of getting this moving forward.

We should also capitalize that class and merge the contents from the jenkins-utils into the parser class.

Dhivyaa21 commented:
@cdancy  - Removing the trailing forward slash is a good idea. I have modified the code, Please review.

cdancy commented on src/main/java/com/cdancy/jenkins/rest/parsers/OptionalFolderPathParser.java:
```diff
@@ -0,0 +1,32 @@
+        String folderPath = String.class.cast(optionalFolderPath);
+        String amendedPath = "";
+        if(folderPath.startsWith("/") || folderPath.endsWith("/")) {
+            throw new RuntimeException("Incorrect folder Path format - " + folderPath);
```
I'm actually going to push back on this, and go against what I originally said, and say we allow both. If either are present just remove before processing. I can see someone doing something like: `/hello/world` and someone else looking at the api, not seeing a forward slash, and then do something like: `hello/world/`.

cdancy commented on src/main/java/com/cdancy/jenkins/rest/parsers/OptionalFolderPathParser.java:
```diff
@@ -0,0 +1,32 @@
+        }
+        String[] folderNames = folderPath.split("/");
+        for (String folder:folderNames) {
+            amendedPath += "job/" + folder + "/";
```
I wonder about this: what happens if the user puts in a non-simple path? Meaning the fully qualified `job/hello/job/world/job/build`.

Dhivyaa21 replied:
Like @martinda pointed, It is hard to say if "job/job" is a fully qualified path to a job or simple folder path(Apparently, folders can also be named as "job"). I am not sure if we can check for a fully qualified path before we amend it.

When users say `job/hello/job/world/job/build`  this logic will go ahead add `job/` before every word and probably would return a 404 not found status.

martinda replied:
In the fully qualified job name `job/hello/job/world/job/build`, there are two things to consider:
1. The path (hello/world) - forward slashes used for nesting if necessary.
1. The job name (build) - forward slashes not allowed.

This pattern reminds me of the Java [`File(String parent, String child)`](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#File-java.lang.String-java.lang.String-) constructor, for which there are effectively multiple constructors. It also reminds me of the [`createDirectories`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#createDirectories-java.nio.file.Path-java.nio.file.attribute.FileAttribute...-) in nio Files.

Let's imagine for a moment that there is an API to create a folder path, and that it returns a `JenkinsFolder` instance:

```java
    JenkinsFolder createFolders(String folderName)
```
That api would create folders on Jenkins, thereby validating the user input (simple path expression) and returning some read-only instance representing that path.

Then there would be two APIs to create a job:
1. `create(JenkinsFolder folder, String jobName)`
2. `create(String jobName)`

Regarding the path, the regex to detect a fully qualified path is: `Pattern pattern = Pattern.compile('(job/[^/]+?)(/job/[^/])*');`. However it fails if the user is "crazy" enough to call a folder "job" and a nested folder "job", because in this case it is ambiguous.

Ultimately, only Jenkins can completely validate the `folderPath` and the `jobName`. IMO it is fine to let Jenkins return 404 if the user incorrectly specifies them. I would prefer that jenkins-rest does not interfere with what Jenkins can accept.

I also prefer a solution that is predictable from the user point of view. If the internals contain different behaviors that depend on the input value, those need to be explained in the wiki. Potentially, the more internal manipulations the jenkins-rest does, the more difficult it is for the user to debug.

I think a lot... perhaps too much... what do you think?

martinda replied:
I have not considered all the cases. Specifically, the case where a `JenkinsFolder` already exists. In this case, `createFolder(String folerName)` is not the right call to create an instance of `JenkinsFolder`.

cdancy replied:
@martinda so is the thought here that users can ONLY pass in the "simple names"? Would it make sense to check each iteration and if it's not preceded with "job" then insert it otherwise assume OK? I'd like to support both the simple case and the normal case if possible.

martinda replied:
Let's assume:
1. simple names, simple case: `a/b/c`
1. normal names, normal case: `job/a/job/b/job/c` (aka fully qualified folder)

> so is the thought here that users can ONLY pass in the "simple names"?

Yes.

> Would it make sense to check each iteration and if it's not preceded with "job" then insert it otherwise assume OK?

Yes but there is an ambiguity: a user can create a folder called "job". Unfortunately, this is possible (I tried it). Because of the ambiguity, the code will not work if a user creates nested folders called "job" (this would be strange, but Jenkins allows it).

>I'd like to support both the simple case and the normal case if possible.

Me too. But I do not know what to do about the ambiguity. So in my mind the code should only support one of the two cases. I am good with either one.

cdancy replied:
@martinda while I though there must be _something_ we can do, after playing with things for a while, I see that its impossible given that a job name can actually be `job`. If it weren't for this we'd be able to check but it's simply impossible to code for things if `job` is a valid name.

cdancy commented on src/main/java/com/cdancy/jenkins/rest/parsers/OptionalFolderPathParser.java:
```diff
@@ -20,9 +20,15 @@ public String apply(Object optionalFolderPath) {
+        if(folderPath.endsWith("/")) {
+            folderPath = folderPath.replaceAll("/+$","");
         }
+
```
@Dhivyaa21 looks good. With all the string manipulation would it be better to use a StringBuilder here? It would remove the need to use regex's via the `replaceAll` method and instead we could just remove the first and last character if present. Would also speed up the concatenation of strings below.

cdancy commented on src/main/java/com/cdancy/jenkins/rest/parsers/OptionalFolderPathParser.java:
```diff
@@ -0,0 +1,38 @@
+            path.deleteCharAt(path.length()-1);
+        }
+
+        String[] folders = path.toString().split("/");
```
Think we can get away with 1 StringBuilder here. Once we do this split we can zero out the `path` builder and just append to that.

cdancy commented on src/main/java/com/cdancy/jenkins/rest/parsers/OptionalFolderPathParser.java:
```diff
@@ -0,0 +1,38 @@
+
+        String[] folders = path.toString().split("/");
+        for(String folder:folders) {
+            amendedPath.append("job/").append(folder).append("/");
```
Lets make these constants within this class as well.

Dhivyaa21 replied:
did you mean the "job/" should be static final String?

cdancy replied:
Exactly. That and references to `"/"`. Basically anything we're creating more than once and will potentially be used multiple times through the lifecycle of this library being used. Java will most likely cache this correctly behind the scenes but still: I like to aggressively do so within my code ;)

cdancy commented:
Pulling this in now. Thanks for the help everyone!

cdancy commented:
Version `0.0.11` has been released.
